### PR TITLE
ElasticSearch Processing

### PR DESCRIPTION
**Pull request by ahurka**
Adds new ElasticSearch indexes and attaches them to processing routine. Adds ElasticSearch handling to model CLI. Introduces the `admin search sync` command, which (attempts to) precisely sync the search index with the registry.

### CODE REVIEW

File: woudc_data_registry/epicentre/metadata.py [hunk 1]
```diff
@@ -46,12 +46,14 @@
 from datetime import date
 import logging
 
-from woudc_data_registry import registry
-from woudc_data_registry.models import Contributor, Country, StationName
+from woudc_data_registry import registry, search
+from woudc_data_registry.models import Contributor, Country, \
+                                       Station, StationName
```
Comment (tomkralidis) [inline woudc_data_registry/epicentre/metadata.py]: better written as:
```python
from woudc_data_registry.models import (Contributor, Country,
    Station, StationName)

File: woudc_data_registry/epicentre/metadata.py [hunk 2]
```diff
@@ -128,15 +130,18 @@ def add_metadata(entity, dict_):
...
+    if psql:
+        REGISTRY.save(c)
+    if es:
+        SEARCH_INDEX.index(entity, c.__geo_interface__)
 
     return c
 
 
-def update_metadata(entity, identifier, dict_):
+def update_metadata(entity, identifier, dict_, psql=True, es=True):
```
Comment (tomkralidis) [inline woudc_data_registry/epicentre/metadata.py]: What's the goal of `psql` flags here?

File: woudc_data_registry/models.py [hunk 3]
```diff
@@ -602,15 +616,15 @@ def __init__(self, ecsv):
             self.data_generation_scientific_authority = \
                 ecsv.extcsv['DATA_GENERATION']['ScientificAuthority']
 
-        self.platform_type = ecsv.extcsv['PLATFORM']['Type']
-        self.platform_name = ecsv.extcsv['PLATFORM']['Name']
-        self.platform_country = ecsv.extcsv['PLATFORM']['Country']
-        self.platform_gaw_id = ecsv.extcsv['PLATFORM'].get('GAW_ID', None)
+        self._platform_type = ecsv.extcsv['PLATFORM']['Type']
```
Comment (tomkralidis) [inline woudc_data_registry/models.py]: Why are we prefixing all properties with `_` ?

File: woudc_data_registry/processing.py [hunk 4]
```diff
@@ -1020,9 +1030,9 @@ def check_data_record(self):
         version = self.extcsv.extcsv['DATA_GENERATION']['Version']
         dg_valueline = self.extcsv.line_num('DATA_GENERATION') + 2
 
-        components = self.data_record.data_record_id.split(':')
-        components[-1] = '%.%'
-        identifier_pattern = ':'.join(components)
+        id_components = data_record.data_record_id.split(':')
+        id_components[-1] = '%.%'
```
Comment (tomkralidis) [inline woudc_data_registry/processing.py]: What's `%.%` ?

File: woudc_data_registry/models.py (hunk 3 above)
Comment (ahurka) [reply woudc_data_registry/models.py]: Any field with an underscore is also an `@property` down below. Externally you can and should access the station type using the `platform_type` property, so the interface is the same. Internally it either comes from `_platform_type` (if the object was made from an ecsv file and __init__ is called) or from a relationship reference (if the object comes straight from PSQL). That way you can reference `platform_type` no matter which source the object comes from.

File: woudc_data_registry/processing.py (hunk 4 above)
Comment (ahurka) [reply woudc_data_registry/processing.py]: `%.%` is PSQL for the regular expression `".\.."`. Two wildcards separated by a dot. It matches any floating-point version number, and so you end up getting a previously existing version of the data record.

File: woudc_data_registry/epicentre/metadata.py (hunk 2 above)
Comment (ahurka) [reply woudc_data_registry/epicentre/metadata.py]: I've been thinking about adding config options to turn on/off certain relations in the registry or indexes in ES under certain conditions. The `es=True` flag had to be added anyways, so it made sense to add both these flags in at once even if one is unused.

File: woudc_data_registry/processing.py (hunk 4 above)
Comment (tomkralidis) [reply woudc_data_registry/processing.py]: can we make `registry.query_by_pattern` do this logic?  Having `%.%` which is an SQL-ism is better isolated in `woudc_data_registry.registry.query_by_pattern`.

**Pull request merged**